internal/characters/barbara/abil.go

```diff
@@ -0,0 +1,197 @@
+package barbara
+
+import (
+	"fmt"
+
+	"github.com/genshinsim/gcsim/pkg/core"
+)
+
+// Standard attack function with seal handling
+func (c *char) Attack(p map[string]int) (int, int) {
+	f, a := c.ActionFrames(core.ActionAttack, p)
+
+	ai := core.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       fmt.Sprintf("Normal %v", c.NormalCounter),
+		AttackTag:  core.AttackTagNormal,
+		ICDTag:     core.ICDTagNormalAttack,
+		ICDGroup:   core.ICDGroupDefault,
+		Element:    core.Hydro,
+		Durability: 25,
+		Mult:       attack[c.NormalCounter][c.TalentLvlAttack()],
+	}
+	done := false
+	// Taken from Noelle code
+	cb := func(a core.AttackCB) {
+		if done { //why do we need this @srl
+			return
+		}
+		//check for healing
+		if c.Core.Status.Duration("barbskill") > 0 {
+			//heal target
+			heal := (prochpp[c.TalentLvlSkill()] + prochp[c.TalentLvlSkill()])
+			c.Core.Health.HealAll(c.Index, heal)
+			done = true
+		}
+
+	}
+	c.Core.Combat.QueueAttack(ai, core.NewDefCircHit(0.1, false, core.TargettableEnemy), 0, f, cb)
+	c.AdvanceNormalIndex()
+
+	// return animation cd
+	return f, a
+}
+
+// Charge attack function - handles seal use
+func (c *char) ChargeAttack(p map[string]int) (int, int) {
+
+	f, a := c.ActionFrames(core.ActionCharge, p)
+
+	ai := core.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       "Charge Attack",
+		AttackTag:  core.AttackTagExtra,
+		ICDTag:     core.ICDTagNone,
+		ICDGroup:   core.ICDGroupDefault,
+		StrikeType: core.StrikeTypeBlunt,
+		Element:    core.Hydro,
+		Durability: 25,
+		Mult:       charge[c.NormalCounter][c.TalentLvlAttack()],
+	}
+
+	done := false
+	// Taken from Noelle code
+	cb := func(a core.AttackCB) {
+		if done { //why do we need this @srl
+			return
+		}
+		//check for healing
+		if c.Core.Status.Duration("barbskill") > 0 {
+			//heal target
+			heal := (prochpp[c.TalentLvlSkill()] + prochp[c.TalentLvlSkill()])
+			c.Core.Health.HealAll(c.Index, 4*heal)
+			done = true
+		}
+
+	}
+	var cbenergy func(a core.AttackCB) = nil
+	energyCount := 0
```

Comment:
this should prob be moved into the if statement so there's no wasted mem alloc if cons < 4?